{
  "cells": [
    {
      "cell_type": "markdown",
      "id": "2f8f89a9",
      "metadata": {
        "id": "2f8f89a9"
      },
      "source": [
        "\n",
        "# üß¨ CRISPR Base Editing Practical with BEstimate\n",
        "\n",
        "Welcome! This practical will walk you through:\n",
        "\n",
        "‚úÖ Preparing your environment (in Colab)  \n",
        "\n",
        "‚úÖ Preparing the input and running BEstimate\n",
        "\n",
        "‚úÖ Interpreting the results\n",
        "\n",
        "‚úÖ Library design  \n",
        "\n",
        "---\n"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "4b841fd7",
      "metadata": {
        "id": "4b841fd7"
      },
      "source": [
        "\n",
        "## 1Ô∏è‚É£ Pre-Preparation (Run This Before the Practical)\n",
        "\n",
        "> **Important:** Run these two sections to install everything you will need for the training.\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 3,
      "id": "8d047ec7",
      "metadata": {
        "id": "8d047ec7",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "collapsed": true,
        "outputId": "1b6b7aeb-407c-4cf0-c516-230e6f421637"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Cloning into 'BEstimate'...\n",
            "remote: Enumerating objects: 826, done.\u001b[K\n",
            "remote: Counting objects: 100% (30/30), done.\u001b[K\n",
            "remote: Compressing objects: 100% (23/23), done.\u001b[K\n",
            "remote: Total 826 (delta 11), reused 16 (delta 7), pack-reused 796 (from 2)\u001b[K\n",
            "Receiving objects: 100% (826/826), 5.96 MiB | 16.71 MiB/s, done.\n",
            "Resolving deltas: 100% (466/466), done.\n"
          ]
        }
      ],
      "source": [
        "import os\n",
        "\n",
        "# Install BEstimate directly from GitHub\n",
        "!git clone https://github.com/CansuDincer/BEstimate.git\n",
        "os.chdir(\"/content/BEstimate/\")\n"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "!pip3 install -r /content/BEstimate/requirements.txt"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "RRPiT6hbjLCw",
        "outputId": "550086ee-6489-4b43-c270-af52b65a90c2"
      },
      "id": "RRPiT6hbjLCw",
      "execution_count": 4,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Collecting pandas==2.2.3 (from -r /content/BEstimate/requirements.txt (line 1))\n",
            "  Downloading pandas-2.2.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (89 kB)\n",
            "\u001b[2K     \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m89.9/89.9 kB\u001b[0m \u001b[31m3.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting argparse==1.1 (from -r /content/BEstimate/requirements.txt (line 2))\n",
            "  Downloading argparse-1.1.zip (151 kB)\n",
            "\u001b[2K     \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m151.5/151.5 kB\u001b[0m \u001b[31m7.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: requests==2.32.3 in /usr/local/lib/python3.11/dist-packages (from -r /content/BEstimate/requirements.txt (line 3)) (2.32.3)\n",
            "Collecting numpy==2.2.5 (from -r /content/BEstimate/requirements.txt (line 4))\n",
            "  Downloading numpy-2.2.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (62 kB)\n",
            "\u001b[2K     \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m62.0/62.0 kB\u001b[0m \u001b[31m4.4 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hCollecting biopython==1.85 (from -r /content/BEstimate/requirements.txt (line 5))\n",
            "  Downloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (13 kB)\n",
            "Requirement already satisfied: multiprocess in /usr/local/lib/python3.11/dist-packages (from -r /content/BEstimate/requirements.txt (line 6)) (0.70.15)\n",
            "Collecting getopts (from -r /content/BEstimate/requirements.txt (line 7))\n",
            "  Downloading getopts-1.0.5.tar.gz (8.4 kB)\n",
            "  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Collecting supyr-struct (from -r /content/BEstimate/requirements.txt (line 8))\n",
            "  Downloading supyr_struct-1.5.4-py3-none-any.whl.metadata (4.7 kB)\n",
            "Collecting python-collection (from -r /content/BEstimate/requirements.txt (line 9))\n",
            "  Downloading python_collection-1.0.0-py3-none-any.whl.metadata (4.8 kB)\n",
            "Collecting numba==0.61.2 (from -r /content/BEstimate/requirements.txt (line 10))\n",
            "  Downloading numba-0.61.2-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (2.8 kB)\n",
            "Collecting dataclasses (from -r /content/BEstimate/requirements.txt (line 11))\n",
            "  Downloading dataclasses-0.6-py3-none-any.whl.metadata (3.0 kB)\n",
            "Collecting typing (from -r /content/BEstimate/requirements.txt (line 12))\n",
            "  Downloading typing-3.7.4.3.tar.gz (78 kB)\n",
            "\u001b[2K     \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m78.6/78.6 kB\u001b[0m \u001b[31m5.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25h  Preparing metadata (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "Requirement already satisfied: python-dateutil>=2.8.2 in /usr/local/lib/python3.11/dist-packages (from pandas==2.2.3->-r /content/BEstimate/requirements.txt (line 1)) (2.9.0.post0)\n",
            "Requirement already satisfied: pytz>=2020.1 in /usr/local/lib/python3.11/dist-packages (from pandas==2.2.3->-r /content/BEstimate/requirements.txt (line 1)) (2025.2)\n",
            "Requirement already satisfied: tzdata>=2022.7 in /usr/local/lib/python3.11/dist-packages (from pandas==2.2.3->-r /content/BEstimate/requirements.txt (line 1)) (2025.2)\n",
            "Requirement already satisfied: charset-normalizer<4,>=2 in /usr/local/lib/python3.11/dist-packages (from requests==2.32.3->-r /content/BEstimate/requirements.txt (line 3)) (3.4.2)\n",
            "Requirement already satisfied: idna<4,>=2.5 in /usr/local/lib/python3.11/dist-packages (from requests==2.32.3->-r /content/BEstimate/requirements.txt (line 3)) (3.10)\n",
            "Requirement already satisfied: urllib3<3,>=1.21.1 in /usr/local/lib/python3.11/dist-packages (from requests==2.32.3->-r /content/BEstimate/requirements.txt (line 3)) (2.4.0)\n",
            "Requirement already satisfied: certifi>=2017.4.17 in /usr/local/lib/python3.11/dist-packages (from requests==2.32.3->-r /content/BEstimate/requirements.txt (line 3)) (2025.7.9)\n",
            "Collecting llvmlite<0.45,>=0.44.0dev0 (from numba==0.61.2->-r /content/BEstimate/requirements.txt (line 10))\n",
            "  Downloading llvmlite-0.44.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (4.8 kB)\n",
            "Requirement already satisfied: dill>=0.3.7 in /usr/local/lib/python3.11/dist-packages (from multiprocess->-r /content/BEstimate/requirements.txt (line 6)) (0.3.7)\n",
            "Requirement already satisfied: six>=1.5 in /usr/local/lib/python3.11/dist-packages (from python-dateutil>=2.8.2->pandas==2.2.3->-r /content/BEstimate/requirements.txt (line 1)) (1.17.0)\n",
            "Downloading pandas-2.2.3-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.1 MB)\n",
            "\u001b[2K   \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m13.1/13.1 MB\u001b[0m \u001b[31m43.3 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading numpy-2.2.5-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (16.4 MB)\n",
            "\u001b[2K   \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m16.4/16.4 MB\u001b[0m \u001b[31m74.5 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading biopython-1.85-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.3 MB)\n",
            "\u001b[2K   \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m3.3/3.3 MB\u001b[0m \u001b[31m76.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading numba-0.61.2-cp311-cp311-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (3.8 MB)\n",
            "\u001b[2K   \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m3.8/3.8 MB\u001b[0m \u001b[31m76.6 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading supyr_struct-1.5.4-py3-none-any.whl (236 kB)\n",
            "\u001b[2K   \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m236.5/236.5 kB\u001b[0m \u001b[31m18.2 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hDownloading python_collection-1.0.0-py3-none-any.whl (3.5 kB)\n",
            "Downloading dataclasses-0.6-py3-none-any.whl (14 kB)\n",
            "Downloading llvmlite-0.44.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (42.4 MB)\n",
            "\u001b[2K   \u001b[90m‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ‚îÅ\u001b[0m \u001b[32m42.4/42.4 MB\u001b[0m \u001b[31m16.1 MB/s\u001b[0m eta \u001b[36m0:00:00\u001b[0m\n",
            "\u001b[?25hBuilding wheels for collected packages: argparse, getopts, typing\n",
            "  Building wheel for argparse (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for argparse: filename=argparse-1.1-py3-none-any.whl size=20717 sha256=e6d2d5196ccf0f9e0ff7eb50d07146fed362a896df854acd2c3c760d17649bd2\n",
            "  Stored in directory: /root/.cache/pip/wheels/bf/26/ab/7a0cfffb3f738c5f058e100d6ecad7be482a14889c7eca6db8\n",
            "  Building wheel for getopts (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for getopts: filename=getopts-1.0.5-py3-none-any.whl size=8486 sha256=793319f82c3865dd88b9e4dd61571cdb76ee7596955d32c15d5127bdef94ee96\n",
            "  Stored in directory: /root/.cache/pip/wheels/42/95/48/da99b31f9278a18a101edca7b4fe3734b381d0d8f92373f2eb\n",
            "  Building wheel for typing (setup.py) ... \u001b[?25l\u001b[?25hdone\n",
            "  Created wheel for typing: filename=typing-3.7.4.3-py3-none-any.whl size=26304 sha256=99a4103aa1d6ea92e750f6a44e9a735fcbd0df82811bb88131d76685a22a321a\n",
            "  Stored in directory: /root/.cache/pip/wheels/9d/67/2f/53e3ef32ec48d11d7d60245255e2d71e908201d20c880c08ee\n",
            "Successfully built argparse getopts typing\n",
            "Installing collected packages: python-collection, getopts, dataclasses, argparse, typing, supyr-struct, numpy, llvmlite, pandas, numba, biopython\n",
            "  Attempting uninstall: numpy\n",
            "    Found existing installation: numpy 2.0.2\n",
            "    Uninstalling numpy-2.0.2:\n",
            "      Successfully uninstalled numpy-2.0.2\n",
            "  Attempting uninstall: llvmlite\n",
            "    Found existing installation: llvmlite 0.43.0\n",
            "    Uninstalling llvmlite-0.43.0:\n",
            "      Successfully uninstalled llvmlite-0.43.0\n",
            "  Attempting uninstall: pandas\n",
            "    Found existing installation: pandas 2.2.2\n",
            "    Uninstalling pandas-2.2.2:\n",
            "      Successfully uninstalled pandas-2.2.2\n",
            "  Attempting uninstall: numba\n",
            "    Found existing installation: numba 0.60.0\n",
            "    Uninstalling numba-0.60.0:\n",
            "      Successfully uninstalled numba-0.60.0\n",
            "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
            "google-colab 1.0.0 requires pandas==2.2.2, but you have pandas 2.2.3 which is incompatible.\n",
            "cuml-cu12 25.2.1 requires numba<0.61.0a0,>=0.59.1, but you have numba 0.61.2 which is incompatible.\n",
            "dask-cuda 25.2.0 requires numba<0.61.0a0,>=0.59.1, but you have numba 0.61.2 which is incompatible.\n",
            "distributed-ucxx-cu12 0.42.0 requires numba<0.61.0a0,>=0.59.1, but you have numba 0.61.2 which is incompatible.\n",
            "tensorflow 2.18.0 requires numpy<2.1.0,>=1.26.0, but you have numpy 2.2.5 which is incompatible.\n",
            "cudf-cu12 25.2.1 requires numba<0.61.0a0,>=0.59.1, but you have numba 0.61.2 which is incompatible.\u001b[0m\u001b[31m\n",
            "\u001b[0mSuccessfully installed argparse-1.1 biopython-1.85 dataclasses-0.6 getopts-1.0.5 llvmlite-0.44.0 numba-0.61.2 numpy-2.2.5 pandas-2.2.3 python-collection-1.0.0 supyr-struct-1.5.4 typing-3.7.4.3\n"
          ]
        },
        {
          "output_type": "display_data",
          "data": {
            "application/vnd.colab-display-data+json": {
              "pip_warning": {
                "packages": [
                  "argparse",
                  "dataclasses",
                  "numpy",
                  "pandas",
                  "typing"
                ]
              },
              "id": "3b19eda6dc074d11813b4571e5ffbcfd"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Please restart your session so the packages will be installed!**"
      ],
      "metadata": {
        "id": "JVQsENgmiIxk"
      },
      "id": "JVQsENgmiIxk"
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Important, I will add the index file separately since downloading and indexing the genome in the practical session are infeasible. However you can use the code as below (x_genome.py) on your own in a linux environment**"
      ],
      "metadata": {
        "id": "x1WdoWlviPfa"
      },
      "id": "x1WdoWlviPfa"
    },
    {
      "cell_type": "code",
      "source": [
        "import os, pandas"
      ],
      "metadata": {
        "id": "XnRTKHvrcBsL"
      },
      "id": "XnRTKHvrcBsL",
      "execution_count": 1,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Make an output folder inside content directory\n",
        "#os.mkdir(\"/content/output/\")\n",
        "os.mkdir(\"/content/BEstimate/offtargets/\")\n",
        "\n",
        "# Change the path to inside BEstimate folder\n",
        "os.chdir(\"/content/BEstimate/BEstimate/\")"
      ],
      "metadata": {
        "id": "tMtAgF7Nb9Jx"
      },
      "id": "tMtAgF7Nb9Jx",
      "execution_count": 3,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#!python3 x_genome.py --pamseq NGG --assembly GRCh38 --ensembl_version 113"
      ],
      "metadata": {
        "id": "wykDux4jiO0m"
      },
      "id": "wykDux4jiO0m",
      "execution_count": 50,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 2Ô∏è‚É£ Designing gRNAs for Base Editors"
      ],
      "metadata": {
        "id": "EJCtrtN8pFeK"
      },
      "id": "EJCtrtN8pFeK"
    },
    {
      "cell_type": "markdown",
      "source": [
        "üß¨ To find the most appropriate gRNA for our experiments, we should decide:\n",
        "\n",
        "1. Length of the protospacer and PAM sequences\n",
        "  - Typically the protospacer sequence is 20.\n",
        "  - PAM is more divergent however the most frequently used one is NGG or NGN\n",
        "2. The sequence interval of the activity window.\n",
        "  - Typically activity window reside between 4-8 or 3-9 nucleotides on the protospacer sequence.\n",
        "3. The editable nucleotides\n",
        "  - CBE or ABE\n",
        "  - For a novel base editor, you can specify any nucleotide change\n",
        "\n",
        "After the information related to Base Editors, you should also decide which gene is your interest:\n",
        "\n",
        "1. Hugo symbol of the gene\n",
        "2. (Optionally) Ensembl Transcript ID\n",
        "3. (Optionally) Uniprot ID\n",
        "4. Any variants you want to incorporate (HGVS structure)\n"
      ],
      "metadata": {
        "id": "37-_tOmoo_Hn"
      },
      "id": "37-_tOmoo_Hn"
    },
    {
      "cell_type": "markdown",
      "id": "20c0ef88",
      "metadata": {
        "id": "20c0ef88"
      },
      "source": [
        "\n",
        "##3Ô∏è‚É£ Running BEstimate on Example Genes\n",
        "\n",
        "Let's design base editor guides for *SRY* as practice.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Mutagenesis on *SRY* gene**"
      ],
      "metadata": {
        "id": "kotm4Xftsl_w"
      },
      "id": "kotm4Xftsl_w"
    },
    {
      "cell_type": "code",
      "execution_count": 17,
      "id": "725d7549",
      "metadata": {
        "id": "725d7549",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "79961fa7-013b-4e12-f2cf-fbca89e1ba3b"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
            "  warnings.warn(\n",
            "\n",
            "--------------------------------------------------------------                                                                                         \n",
            "\t\t   B E s t i m a t e                                      \n",
            "\n",
            "\t       Wellcome Sanger Institute          \n",
            "\n",
            "--------------------------------------------------------------\n",
            "    \n",
            "\n",
            "The given arguments are:\n",
            "Gene: SRY\n",
            "Assembl: GRCh38\n",
            "Ensembl transcript ID: None\n",
            "Uniprot ID: None\n",
            "PAM sequence: NGG\n",
            "PAM window: 21-23\n",
            "Protospacer length: 20\n",
            "Activity window: 4-8\n",
            "Nucleotide change: A>G\n",
            "VEP and Uniprot analysis: True\n",
            "Mutation on genome: \n",
            "Off target analysis: False\n",
            "\n",
            "\n",
            "\n",
            "-------------------------------------------------------------- \n",
            "\t\tEnsembl Gene Information\n",
            "-------------------------------------------------------------- \n",
            "    \n",
            "\n",
            "Request to Ensembl REST API for Ensembl Gene ID:\n",
            ">ENSG00000184895.8 chromosome:GRCh38:Y:2786855:2787682:-1\n",
            "Ensembl Gene ID: ENSG00000184895\n",
            "\n",
            "Request to Ensembl REST API for sequence information\n",
            "The location of the interested gene: chromosome:GRCh38:Y:2786855:2787682:-1\n",
            "\n",
            "Sequence was retrieved.\n",
            "\n",
            "\n",
            "--------------------------------------------------------------\n",
            "\t\tgRNAs - Targetable Sites\n",
            "--------------------------------------------------------------\n",
            "    \n",
            "\n",
            "CRISPR Data Frame was read from /content/output/ as SRY_ABE_NGG_crispr_df.csv\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "--------------------------------------------------------------\n",
            "\t\tgRNAs - Editable Sites\n",
            "--------------------------------------------------------------\n",
            "    \n",
            "\n",
            "Edit Data Frame was read from /content/output/ as SRY_ABE_NGG_edit_df.csv\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "--------------------------------------------------------------\n",
            "\t\tAnnotation - VEP Annotation\n",
            "--------------------------------------------------------------\n",
            "        \n",
            "\n",
            "HGVS nomenclatures were collected.\n",
            "\n",
            "VEP Data Frame is filling with VEP API.\n",
            "VEP Data Frame was created!\n",
            "VEP Data Frame was written in /content/output/ as SRY_ABE_NGG_vep_df.csv\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "--------------------------------------------------------------\n",
            "\t\tAnnotation - Protein Annotation\n",
            "--------------------------------------------------------------\n",
            "\t\t\n",
            "\n",
            "Adding Uniprot ID, corresponding Domain and PTM information..\n",
            "Uniprot ID: Q05066\n",
            "Adding affected interface and interacting partners..\n",
            "Protein Data Frame was created!\n",
            "Protein Data Frame was written in /content/output/ as SRY_ABE_NGG_protein_df.csv\n",
            "\n",
            "\n",
            "Summarising information..\n",
            "Summary Data Frame was created!\n",
            "Summary Data Frame was written in /content/output/ as SRY_ABE_NGG_summary_df.csv\n",
            "\n",
            "\n",
            "\n",
            "\n",
            "--------------------------------------------------------------\n",
            "\tThe BEstimate analysis successfully finished!\n",
            "--------------------------------------------------------------\n",
            "\t\n",
            "\n"
          ]
        }
      ],
      "source": [
        "# Run BEstimate with example input\n",
        "!python3 BEstimate.py -gene SRY -assembly GRCh38 -pamseq NGG -pamwin 21-23 -actwin 4-8 -protolen 20 -edit A -edit_to G -vep -o /content/output/ -ofile SRY_ABE_NGG\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### **Reverting sickle-cel disease associated variant**"
      ],
      "metadata": {
        "id": "3J8aHR9tssXD"
      },
      "id": "3J8aHR9tssXD"
    },
    {
      "cell_type": "markdown",
      "source": [
        "Sickle cell disease - mutation Œ≤-globin gene (*HBB*): g.5227002A>T in GRCh38, p.Glu7Val"
      ],
      "metadata": {
        "id": "Y02QtvFztR50"
      },
      "id": "Y02QtvFztR50"
    },
    {
      "cell_type": "code",
      "source": [
        "# Generate a mutation file\n",
        "f = open(\"/content/sicle_cell_variant.txt\", \"w\")\n",
        "f.writelines(\"11:g.5227002A>T\")\n",
        "f.close()"
      ],
      "metadata": {
        "id": "C95d-ZiptiRO"
      },
      "id": "C95d-ZiptiRO",
      "execution_count": 4,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Run BEstimate with example input\n",
        "!python3 BEstimate.py -gene HBB -assembly GRCh38 -transcript ENST00000335295 -mutation_file /content/sicle_cell_variant.txt -pamseq NGN -pamwin 21-23 -actwin 3-9 -protolen 20 -edit A -edit_to G -o /content/output/ -ofile HBB_variant_specific_ABE_NGN\n"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-6ziVdEssypM",
        "outputId": "3512f952-ef2f-47ca-d78c-cf979f0ee440"
      },
      "id": "-6ziVdEssypM",
      "execution_count": 15,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "/usr/local/lib/python3.11/dist-packages/Bio/pairwise2.py:278: BiopythonDeprecationWarning: Bio.pairwise2 has been deprecated, and we intend to remove it in a future release of Biopython. As an alternative, please consider using Bio.Align.PairwiseAligner as a replacement, and contact the Biopython developers if you still need the Bio.pairwise2 module.\n",
            "  warnings.warn(\n",
            "\n",
            "--------------------------------------------------------------                                                                                         \n",
            "\t\t   B E s t i m a t e                                      \n",
            "\n",
            "\t       Wellcome Sanger Institute          \n",
            "\n",
            "--------------------------------------------------------------\n",
            "    \n",
            "\n",
            "The given arguments are:\n",
            "Gene: HBB\n",
            "Assembl: GRCh38\n",
            "Ensembl transcript ID: ENST00000335295\n",
            "Uniprot ID: None\n",
            "PAM sequence: NGN\n",
            "PAM window: 21-23\n",
            "Protospacer length: 20\n",
            "Activity window: 3-9\n",
            "Nucleotide change: A>G\n",
            "VEP and Uniprot analysis: False\n",
            "Mutation on genome: 11:g.5227002A>T\n",
            "Off target analysis: False\n",
            "\n",
            "\n",
            "\n",
            "-------------------------------------------------------------- \n",
            "\t\tEnsembl Gene Information\n",
            "-------------------------------------------------------------- \n",
            "    \n",
            "\n",
            "Request to Ensembl REST API for Ensembl Gene ID:\n",
            ">ENSG00000244734.4 chromosome:GRCh38:11:5225464:5229395:-1\n",
            "Ensembl Gene ID: ENSG00000244734\n",
            "\n",
            "Request to Ensembl REST API for sequence information\n",
            "The location of the interested gene: chromosome:GRCh38:11:5225464:5229395:-1\n",
            "\n",
            "11:g.5227002A>T\n",
            "Sequence was retrieved.\n",
            "\n",
            "\n",
            "--------------------------------------------------------------\n",
            "\t\tgRNAs - Targetable Sites\n",
            "--------------------------------------------------------------\n",
            "    \n",
            "\n",
            "Sequence is preparing...\n",
            "Protospacer and PAM regions are searching for right direction...\n",
            "Pattern is created!\n",
            "Pattern is searching through the sequence...\n",
            "CRISPRs are found!\n",
            "\n",
            "Protospacer and PAM regions are searching for left direction...\n",
            "Pattern is created!\n",
            "Pattern is searching through the sequence...\n",
            "CRISPRs are found!\n",
            "\n",
            "CRISPR df is filling...\n",
            "CRISPR Data Frame was created!\n",
            "CRISPR Data Frame was written in /content/output/ as HBB_variant_specific_ABE_NGN_crispr_df.csv\n",
            "\n",
            "\n",
            "\n",
            "--------------------------------------------------------------\n",
            "\t\tgRNAs - Editable Sites\n",
            "--------------------------------------------------------------\n",
            "    \n",
            "\n",
            "Edit Data Frame is filling...\n",
            "Edit Data Frame was created!\n",
            "Edit Data Frame was written in /content/output/ as HBB_variant_specific_ABE_NGN_edit_df.csv\n",
            "\n",
            "\n",
            "\n",
            "--------------------------------------------------------------\n",
            "\tThe BEstimate analysis successfully finished!\n",
            "--------------------------------------------------------------\n",
            "\t\n",
            "\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "id": "2b755f1b",
      "metadata": {
        "id": "2b755f1b"
      },
      "source": [
        "\n",
        "## 4Ô∏è‚É£ Exploring BEstimate Outputs and Interpreting Results\n",
        "\n",
        "Your results are saved in the `/content/output/` folder.\n",
        "\n",
        "To check what was generated, run:\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 6,
      "id": "7040c61b",
      "metadata": {
        "id": "7040c61b",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "12ace7ea-c260-4d37-83ed-385f9c8448f6"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "total 824K\n",
            "-rw-r--r-- 1 root root 201K Jul 15 13:04 HBB_variant_specific_ABE_NGN_crispr_df.csv\n",
            "-rw-r--r-- 1 root root 618K Jul 15 13:05 HBB_variant_specific_ABE_NGN_edit_df.csv\n"
          ]
        }
      ],
      "source": [
        "# List results\n",
        "!ls -lh /content/output/"
      ]
    },
    {
      "cell_type": "markdown",
      "id": "934b5384",
      "metadata": {
        "id": "934b5384"
      },
      "source": [
        "\n",
        "**What to look for:**\n",
        "\n",
        "- Summary `.csv` tables listing guides\n",
        "\n",
        "- Editable nucleotides with annotations of predicted edits\n",
        "\n",
        "\n",
        "You can download these files or open them directly in Colab for inspection.\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "### SRY mutagenesis results"
      ],
      "metadata": {
        "id": "PpI6itIMNF10"
      },
      "id": "PpI6itIMNF10"
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Let's start with the *edit table*, including gRNAs and their editable nucleotides and sequence information**"
      ],
      "metadata": {
        "id": "ql0w6YIkfJxO"
      },
      "id": "ql0w6YIkfJxO"
    },
    {
      "cell_type": "code",
      "source": [
        "edit_df = pandas.read_csv(\"/content/output/SRY_ABE_NGG_edit_df.csv\")\n",
        "edit_df[:5]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 360
        },
        "id": "P7CcwpLIfRzX",
        "outputId": "f60d1f05-1960-4ea4-bbfd-52c00072a557"
      },
      "id": "P7CcwpLIfRzX",
      "execution_count": 52,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "  Hugo_Symbol      CRISPR_PAM_Sequence  gRNA_Target_Sequence  \\\n",
              "0         SRY  GTAAAATAAGTTTCGAACTCTGG  GTAAAATAAGTTTCGAACTC   \n",
              "1         SRY  GTAAAATAAGTTTCGAACTCTGG  GTAAAATAAGTTTCGAACTC   \n",
              "2         SRY  GTAAAATAAGTTTCGAACTCTGG  GTAAAATAAGTTTCGAACTC   \n",
              "3         SRY  GTAAAATAAGTTTCGAACTCTGG  GTAAAATAAGTTTCGAACTC   \n",
              "4         SRY  AAGAGAATATTCCCGCTCTCCGG  AAGAGAATATTCCCGCTCTC   \n",
              "\n",
              "            Location  Edit_Location Direction  Strand          Gene_ID  \\\n",
              "0  Y:2787642-2787664        2787661      left      -1  ENSG00000184895   \n",
              "1  Y:2787642-2787664        2787660      left      -1  ENSG00000184895   \n",
              "2  Y:2787642-2787664        2787659      left      -1  ENSG00000184895   \n",
              "3  Y:2787642-2787664        2787657      left      -1  ENSG00000184895   \n",
              "4  Y:2787517-2787539        2787536      left      -1  ENSG00000184895   \n",
              "\n",
              "     Transcript_ID          Exon_ID  ...  gRNA_flanking_sequences  \\\n",
              "0  ENST00000383070  ENSE00001494622  ...                      NaN   \n",
              "1  ENST00000383070  ENSE00001494622  ...                      NaN   \n",
              "2  ENST00000383070  ENSE00001494622  ...                      NaN   \n",
              "3  ENST00000383070  ENSE00001494622  ...                      NaN   \n",
              "4  ENST00000383070  ENSE00001494622  ...                      NaN   \n",
              "\n",
              "   Edit_in_Exon  Edit_in_CDS   GC%  # Edits/guide  Poly_T  mutation_on_guide  \\\n",
              "0          True        False  30.0              4   False              False   \n",
              "1          True        False  30.0              4   False              False   \n",
              "2          True        False  30.0              4   False              False   \n",
              "3          True        False  30.0              4   False              False   \n",
              "4          True         True  45.0              3   False              False   \n",
              "\n",
              "   guide_change_mutation  mutation_on_window  mutation_on_PAM  \n",
              "0                  False               False            False  \n",
              "1                  False               False            False  \n",
              "2                  False               False            False  \n",
              "3                  False               False            False  \n",
              "4                  False               False            False  \n",
              "\n",
              "[5 rows x 21 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-b54bb7cc-05e2-432a-a930-e2bdd25a5b02\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Hugo_Symbol</th>\n",
              "      <th>CRISPR_PAM_Sequence</th>\n",
              "      <th>gRNA_Target_Sequence</th>\n",
              "      <th>Location</th>\n",
              "      <th>Edit_Location</th>\n",
              "      <th>Direction</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>Transcript_ID</th>\n",
              "      <th>Exon_ID</th>\n",
              "      <th>...</th>\n",
              "      <th>gRNA_flanking_sequences</th>\n",
              "      <th>Edit_in_Exon</th>\n",
              "      <th>Edit_in_CDS</th>\n",
              "      <th>GC%</th>\n",
              "      <th># Edits/guide</th>\n",
              "      <th>Poly_T</th>\n",
              "      <th>mutation_on_guide</th>\n",
              "      <th>guide_change_mutation</th>\n",
              "      <th>mutation_on_window</th>\n",
              "      <th>mutation_on_PAM</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GTAAAATAAGTTTCGAACTCTGG</td>\n",
              "      <td>GTAAAATAAGTTTCGAACTC</td>\n",
              "      <td>Y:2787642-2787664</td>\n",
              "      <td>2787661</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000184895</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>ENSE00001494622</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>True</td>\n",
              "      <td>False</td>\n",
              "      <td>30.0</td>\n",
              "      <td>4</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GTAAAATAAGTTTCGAACTCTGG</td>\n",
              "      <td>GTAAAATAAGTTTCGAACTC</td>\n",
              "      <td>Y:2787642-2787664</td>\n",
              "      <td>2787660</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000184895</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>ENSE00001494622</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>True</td>\n",
              "      <td>False</td>\n",
              "      <td>30.0</td>\n",
              "      <td>4</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GTAAAATAAGTTTCGAACTCTGG</td>\n",
              "      <td>GTAAAATAAGTTTCGAACTC</td>\n",
              "      <td>Y:2787642-2787664</td>\n",
              "      <td>2787659</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000184895</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>ENSE00001494622</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>True</td>\n",
              "      <td>False</td>\n",
              "      <td>30.0</td>\n",
              "      <td>4</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GTAAAATAAGTTTCGAACTCTGG</td>\n",
              "      <td>GTAAAATAAGTTTCGAACTC</td>\n",
              "      <td>Y:2787642-2787664</td>\n",
              "      <td>2787657</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000184895</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>ENSE00001494622</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>True</td>\n",
              "      <td>False</td>\n",
              "      <td>30.0</td>\n",
              "      <td>4</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AAGAGAATATTCCCGCTCTCCGG</td>\n",
              "      <td>AAGAGAATATTCCCGCTCTC</td>\n",
              "      <td>Y:2787517-2787539</td>\n",
              "      <td>2787536</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000184895</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>ENSE00001494622</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>45.0</td>\n",
              "      <td>3</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows √ó 21 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-b54bb7cc-05e2-432a-a930-e2bdd25a5b02')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-b54bb7cc-05e2-432a-a930-e2bdd25a5b02 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-b54bb7cc-05e2-432a-a930-e2bdd25a5b02');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-d41505f6-3e55-4178-8993-ca1ddaf21813\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-d41505f6-3e55-4178-8993-ca1ddaf21813')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-d41505f6-3e55-4178-8993-ca1ddaf21813 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe"
            }
          },
          "metadata": {},
          "execution_count": 52
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Check the information with edit file\n",
        "edit_df.columns"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "Sb09cs_BfR1i",
        "outputId": "6403eb29-bdc4-4983-e309-3397386aca4f"
      },
      "id": "Sb09cs_BfR1i",
      "execution_count": 53,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['Hugo_Symbol', 'CRISPR_PAM_Sequence', 'gRNA_Target_Sequence',\n",
              "       'Location', 'Edit_Location', 'Direction', 'Strand', 'Gene_ID',\n",
              "       'Transcript_ID', 'Exon_ID', 'guide_in_CDS', 'gRNA_flanking_sequences',\n",
              "       'Edit_in_Exon', 'Edit_in_CDS', 'GC%', '# Edits/guide', 'Poly_T',\n",
              "       'mutation_on_guide', 'guide_change_mutation', 'mutation_on_window',\n",
              "       'mutation_on_PAM'],\n",
              "      dtype='object')"
            ]
          },
          "metadata": {},
          "execution_count": 53
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# The number of editable gRNAs\n",
        "len(edit_df.CRISPR_PAM_Sequence.unique())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "DssJQXHvfR3j",
        "outputId": "e982cbe3-3e39-4be2-f9ea-0ec565a62168"
      },
      "id": "DssJQXHvfR3j",
      "execution_count": 54,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "62"
            ]
          },
          "metadata": {},
          "execution_count": 54
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Number of gRNAs within the coding sequence\n",
        "len(edit_df[edit_df.guide_in_CDS].CRISPR_PAM_Sequence.unique())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "7t3zxxtofb4d",
        "outputId": "4c8cd622-dfd2-4f73-9278-6c18e1c0ca56"
      },
      "id": "7t3zxxtofb4d",
      "execution_count": 55,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "53"
            ]
          },
          "metadata": {},
          "execution_count": 55
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Number of gRNAs with editable nucleotide within the coding sequence\n",
        "len(edit_df[edit_df.Edit_in_CDS].CRISPR_PAM_Sequence.unique())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "hA8DsQCvfb6c",
        "outputId": "e70c05f6-f991-4acb-e768-41f5d98b5933"
      },
      "id": "hA8DsQCvfb6c",
      "execution_count": 56,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "52"
            ]
          },
          "metadata": {},
          "execution_count": 56
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Number of gRNAs with editable nucleotide within the coding sequence w/out polyT\n",
        "len(edit_df[(edit_df.Edit_in_CDS) & (~edit_df.Poly_T)].CRISPR_PAM_Sequence.unique())"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "EH9Aza01fb9B",
        "outputId": "c8347024-a7c0-4e1e-efe3-f748bbc8f84a"
      },
      "id": "EH9Aza01fb9B",
      "execution_count": 57,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "50"
            ]
          },
          "metadata": {},
          "execution_count": 57
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Let's continue with the *protein table*, including VEP, Uniprot and Interactome Insider annotations**"
      ],
      "metadata": {
        "id": "V1EAmHNmfwRz"
      },
      "id": "V1EAmHNmfwRz"
    },
    {
      "cell_type": "code",
      "source": [
        "protein_df = pandas.read_csv(\"/content/output/SRY_ABE_NGG_protein_df.csv\", index_col=0)\n",
        "protein_df[:5]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 322
        },
        "id": "4VfMQhHwexUF",
        "outputId": "5443f516-c310-48ee-eecd-45cbbca6390b"
      },
      "id": "4VfMQhHwexUF",
      "execution_count": 58,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "           Hugo_Symbol   Edit_Type      CRISPR_PAM_Sequence  \\\n",
              "Unnamed: 0                                                    \n",
              "0                  SRY  individual  AAAATGGCCATTCTTCCAGGAGG   \n",
              "1                  SRY  individual  AAACAGTAAAGGCAACGTCCAGG   \n",
              "2                  SRY  individual  AAACAGTAAAGGCAACGTCCAGG   \n",
              "3                  SRY    multiple  AAACAGTAAAGGCAACGTCCAGG   \n",
              "4                  SRY  individual  AACGGGACCGCTACAGCCACTGG   \n",
              "\n",
              "           CRISPR_PAM_Location  gRNA_Target_Sequence gRNA_Target_Location  \\\n",
              "Unnamed: 0                                                                  \n",
              "0            Y:2787267-2787289  AAAATGGCCATTCTTCCAGG    Y:2787264-2787289   \n",
              "1            Y:2787432-2787454  AAACAGTAAAGGCAACGTCC    Y:2787429-2787454   \n",
              "2            Y:2787432-2787454  AAACAGTAAAGGCAACGTCC    Y:2787429-2787454   \n",
              "3            Y:2787432-2787454  AAACAGTAAAGGCAACGTCC    Y:2787429-2787454   \n",
              "4            Y:2787001-2787023  AACGGGACCGCTACAGCCAC    Y:2786998-2787023   \n",
              "\n",
              "            Total_Edit    Edit_Location Direction    Transcript_ID  ...  \\\n",
              "Unnamed: 0                                                          ...   \n",
              "0                    1          2787286      left  ENST00000383070  ...   \n",
              "1                    2          2787447      left  ENST00000383070  ...   \n",
              "2                    2          2787450      left  ENST00000383070  ...   \n",
              "3                    2  2787447-2787451      left  ENST00000383070  ...   \n",
              "4                    1          2787017      left  ENST00000383070  ...   \n",
              "\n",
              "           Protein_Position  is_disruptive_interface_EXP  \\\n",
              "Unnamed: 0                                                 \n",
              "0                       106                        False   \n",
              "1                        53                        False   \n",
              "2                        52                        False   \n",
              "3                     52;53                        False   \n",
              "4                       196                        False   \n",
              "\n",
              "            is_disruptive_interface_MOD  is_disruptive_interface_PRED  \\\n",
              "Unnamed: 0                                                              \n",
              "0                                 False                         False   \n",
              "1                                 False                         False   \n",
              "2                                 False                         False   \n",
              "3                                 False                         False   \n",
              "4                                 False                         False   \n",
              "\n",
              "            disrupted_PDB_int_partners  disrupted_I3D_int_partners  \\\n",
              "Unnamed: 0                                                           \n",
              "0                                  NaN                         NaN   \n",
              "1                                  NaN                         NaN   \n",
              "2                                  NaN                         NaN   \n",
              "3                                  NaN                         NaN   \n",
              "4                                  NaN                         NaN   \n",
              "\n",
              "            disrupted_Eclair_int_partners  disrupted_PDB_int_genes  \\\n",
              "Unnamed: 0                                                           \n",
              "0                                     NaN                      NaN   \n",
              "1                                     NaN                      NaN   \n",
              "2                                     NaN                      NaN   \n",
              "3                                     NaN                      NaN   \n",
              "4                                     NaN                      NaN   \n",
              "\n",
              "            disrupted_I3D_int_genes  disrupted_Eclair_int_genes  \n",
              "Unnamed: 0                                                       \n",
              "0                               NaN                         NaN  \n",
              "1                               NaN                         NaN  \n",
              "2                               NaN                         NaN  \n",
              "3                               NaN                         NaN  \n",
              "4                               NaN                         NaN  \n",
              "\n",
              "[5 rows x 76 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-32752e44-b657-41e0-8085-76fce3ecb0f5\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Hugo_Symbol</th>\n",
              "      <th>Edit_Type</th>\n",
              "      <th>CRISPR_PAM_Sequence</th>\n",
              "      <th>CRISPR_PAM_Location</th>\n",
              "      <th>gRNA_Target_Sequence</th>\n",
              "      <th>gRNA_Target_Location</th>\n",
              "      <th>Total_Edit</th>\n",
              "      <th>Edit_Location</th>\n",
              "      <th>Direction</th>\n",
              "      <th>Transcript_ID</th>\n",
              "      <th>...</th>\n",
              "      <th>Protein_Position</th>\n",
              "      <th>is_disruptive_interface_EXP</th>\n",
              "      <th>is_disruptive_interface_MOD</th>\n",
              "      <th>is_disruptive_interface_PRED</th>\n",
              "      <th>disrupted_PDB_int_partners</th>\n",
              "      <th>disrupted_I3D_int_partners</th>\n",
              "      <th>disrupted_Eclair_int_partners</th>\n",
              "      <th>disrupted_PDB_int_genes</th>\n",
              "      <th>disrupted_I3D_int_genes</th>\n",
              "      <th>disrupted_Eclair_int_genes</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SRY</td>\n",
              "      <td>individual</td>\n",
              "      <td>AAAATGGCCATTCTTCCAGGAGG</td>\n",
              "      <td>Y:2787267-2787289</td>\n",
              "      <td>AAAATGGCCATTCTTCCAGG</td>\n",
              "      <td>Y:2787264-2787289</td>\n",
              "      <td>1</td>\n",
              "      <td>2787286</td>\n",
              "      <td>left</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>...</td>\n",
              "      <td>106</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SRY</td>\n",
              "      <td>individual</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCCAGG</td>\n",
              "      <td>Y:2787432-2787454</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCC</td>\n",
              "      <td>Y:2787429-2787454</td>\n",
              "      <td>2</td>\n",
              "      <td>2787447</td>\n",
              "      <td>left</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>...</td>\n",
              "      <td>53</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SRY</td>\n",
              "      <td>individual</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCCAGG</td>\n",
              "      <td>Y:2787432-2787454</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCC</td>\n",
              "      <td>Y:2787429-2787454</td>\n",
              "      <td>2</td>\n",
              "      <td>2787450</td>\n",
              "      <td>left</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>...</td>\n",
              "      <td>52</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SRY</td>\n",
              "      <td>multiple</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCCAGG</td>\n",
              "      <td>Y:2787432-2787454</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCC</td>\n",
              "      <td>Y:2787429-2787454</td>\n",
              "      <td>2</td>\n",
              "      <td>2787447-2787451</td>\n",
              "      <td>left</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>...</td>\n",
              "      <td>52;53</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SRY</td>\n",
              "      <td>individual</td>\n",
              "      <td>AACGGGACCGCTACAGCCACTGG</td>\n",
              "      <td>Y:2787001-2787023</td>\n",
              "      <td>AACGGGACCGCTACAGCCAC</td>\n",
              "      <td>Y:2786998-2787023</td>\n",
              "      <td>1</td>\n",
              "      <td>2787017</td>\n",
              "      <td>left</td>\n",
              "      <td>ENST00000383070</td>\n",
              "      <td>...</td>\n",
              "      <td>196</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5 rows √ó 76 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-32752e44-b657-41e0-8085-76fce3ecb0f5')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-32752e44-b657-41e0-8085-76fce3ecb0f5 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-32752e44-b657-41e0-8085-76fce3ecb0f5');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-681795e7-4af6-4877-807b-230aef25ca3e\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-681795e7-4af6-4877-807b-230aef25ca3e')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-681795e7-4af6-4877-807b-230aef25ca3e button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe"
            }
          },
          "metadata": {},
          "execution_count": 58
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Check the information with protein file\n",
        "protein_df.columns"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "MOSTOnWhfu7M",
        "outputId": "e019a0a6-08ad-4d8c-941a-d08f68b49af0"
      },
      "id": "MOSTOnWhfu7M",
      "execution_count": 59,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "Index(['Hugo_Symbol', 'Edit_Type', 'CRISPR_PAM_Sequence',\n",
              "       'CRISPR_PAM_Location', 'gRNA_Target_Sequence', 'gRNA_Target_Location',\n",
              "       'Total_Edit', 'Edit_Location', 'Direction', 'Transcript_ID', 'Exon_ID',\n",
              "       'guide_in_CDS', 'gRNA_flanking_sequences', 'Edit_in_Exon',\n",
              "       'Edit_in_CDS', 'mutation_on_guide', 'guide_change_mutation',\n",
              "       'mutation_on_window', 'mutation_on_PAM', '# Edits/guide', 'Poly_T',\n",
              "       'GC%', 'HGVS', 'Protein_ID', 'VEP_input', 'allele',\n",
              "       'variant_classification', 'most_severe_consequence',\n",
              "       'consequence_terms', 'variant_biotype', 'Regulatory_ID', 'Motif_ID',\n",
              "       'TFs_on_motif', 'cDNA_Change', 'Edited_Codon', 'New_Codon',\n",
              "       'CDS_Position', 'Protein_Position_ensembl', 'Protein_Change',\n",
              "       'Edited_AA', 'Edited_AA_Prop', 'New_AA', 'New_AA_Prop', 'is_Synonymous',\n",
              "       'is_Stop', 'proline_addition', 'swissprot_vep', 'uniprot_provided',\n",
              "       'polyphen_score', 'polyphen_prediction', 'sift_score',\n",
              "       'sift_prediction', 'cadd_phred', 'cadd_raw', 'lof', 'impact',\n",
              "       'blosum62', 'is_clinical', 'clinical_id', 'clinical_significance',\n",
              "       'cosmic_id', 'clinvar_id', 'ancestral_populations', 'Domain',\n",
              "       'curated_Domain', 'PTM', 'Protein_Position',\n",
              "       'is_disruptive_interface_EXP', 'is_disruptive_interface_MOD',\n",
              "       'is_disruptive_interface_PRED', 'disrupted_PDB_int_partners',\n",
              "       'disrupted_I3D_int_partners', 'disrupted_Eclair_int_partners',\n",
              "       'disrupted_PDB_int_genes', 'disrupted_I3D_int_genes',\n",
              "       'disrupted_Eclair_int_genes'],\n",
              "      dtype='object')"
            ]
          },
          "metadata": {},
          "execution_count": 59
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**!!!Since there can be several editable nucleotide, multiple edits on the sequence with a gRNA is possible.**"
      ],
      "metadata": {
        "id": "LnvRuk2wgF78"
      },
      "id": "LnvRuk2wgF78"
    },
    {
      "cell_type": "code",
      "source": [
        "# The most severe consequences from the *SRY* gene targteing gRNAs\n",
        "protein_df.most_severe_consequence.unique()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "djk5-lN8fEeX",
        "outputId": "49aa5afd-8669-4842-f040-018de7096db8"
      },
      "id": "djk5-lN8fEeX",
      "execution_count": 60,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['synonymous_variant', 'missense_variant', '3_prime_UTR_variant',\n",
              "       '5_prime_UTR_variant', 'start_lost', 'stop_lost'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 60
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Protein positions of the potential edits\n",
        "protein_df.Protein_Position.unique()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "fyVxNpZ4gItV",
        "outputId": "19c757d3-07e2-4a6e-c6f4-b87dfe510c38"
      },
      "id": "fyVxNpZ4gItV",
      "execution_count": 61,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['106', '53', '52', '52;53', '196', '160', '24', '23', '23;24',\n",
              "       '155', '161', '160;161', '190', '128', '127', '127;128', nan,\n",
              "       '200', '112', '65', '105', '90', '89', '88', '89;90', '43', '48',\n",
              "       '172', '170', '203', '202', '93', '92', '92;93', '156', '129',\n",
              "       '176', '175', '175;176', '73', '44', '43;44', '102', '34', '137',\n",
              "       '152', '151', '151;152', '40', '69', '68', '68;69', '116', '124',\n",
              "       '191', '110', '109', '109;110', '198', '101', '32', '145', '144',\n",
              "       '144;145', '25', '154', '1'], dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 61
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Targeted functional domains\n",
        "protein_df.curated_Domain.unique()"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "NJxPYco3gIvZ",
        "outputId": "ac9447b9-dd1a-4837-b94d-cceac528a93d"
      },
      "id": "NJxPYco3gIvZ",
      "execution_count": 62,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "array(['Sufficient for interaction with KPNB1', nan, 'Disordered',\n",
              "       'Sufficient for interaction with EP300',\n",
              "       'Necessary for interaction with SLC9A3R2',\n",
              "       'Required for nuclear localization',\n",
              "       'Necessary for interaction with ZNF208 isoform KRAB-O'],\n",
              "      dtype=object)"
            ]
          },
          "metadata": {},
          "execution_count": 62
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Whether any gRNAs with clinical consequences\n",
        "protein_df[~pandas.isna(protein_df.is_clinical) & (protein_df.is_clinical)][[\n",
        "    'Hugo_Symbol', 'gRNA_Target_Sequence', 'most_severe_consequence', 'Edited_AA','New_AA', 'clinical_id']]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 771
        },
        "id": "ALS_JnXsgkuJ",
        "outputId": "6df81a96-45c2-4f58-cff6-140c6a7320a7"
      },
      "id": "ALS_JnXsgkuJ",
      "execution_count": 63,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "           Hugo_Symbol  gRNA_Target_Sequence most_severe_consequence  \\\n",
              "Unnamed: 0                                                             \n",
              "0                  SRY  AAAATGGCCATTCTTCCAGG      synonymous_variant   \n",
              "2                  SRY  AAACAGTAAAGGCAACGTCC        missense_variant   \n",
              "6                  SRY  AAGAGAATATTCCCGCTCTC        missense_variant   \n",
              "15                 SRY  ATTATAAGTATCGACCTCGT        missense_variant   \n",
              "17                 SRY  ATTATAAGTATCGACCTCGT        missense_variant   \n",
              "24                 SRY  CCATGAACGCATTCATCGTG        missense_variant   \n",
              "26                 SRY  CGAAAAATGGCCATTCTTCC      synonymous_variant   \n",
              "27                 SRY  CGAAAAATGGCCATTCTTCC        missense_variant   \n",
              "28                 SRY  CGAAAAATGGCCATTCTTCC        missense_variant   \n",
              "32                 SRY  CTCAGAGATCAGCAAGCAGC        missense_variant   \n",
              "33                 SRY  CTCAGAGATCAGCAAGCAGC      synonymous_variant   \n",
              "41                 SRY  GAAACGGGAGAAAACAGTAA        missense_variant   \n",
              "50                 SRY  GCAAGCAGCTGGGATACCAG        missense_variant   \n",
              "73                 SRY  TCAGAGATCAGCAAGCAGCT        missense_variant   \n",
              "86                 SRY  AAAGGAAGGAAGAGCTTCTC        missense_variant   \n",
              "101                SRY  CACACGATGAATGCGTTCAT        missense_variant   \n",
              "104                SRY  CCACACGATGAATGCGTTCA        missense_variant   \n",
              "106                SRY  CCTGTAATTTCTGTGCCTCC        missense_variant   \n",
              "107                SRY  CCTGTAATTTCTGTGCCTCC      synonymous_variant   \n",
              "113                SRY  CTGGAAGAATGGCCATTTTT        missense_variant   \n",
              "121                SRY  GAGGTCGATACTTATAATTC      synonymous_variant   \n",
              "131                SRY  GGAATATTCTCTTGCACAGC      synonymous_variant   \n",
              "\n",
              "           Edited_AA New_AA   clinical_id  \n",
              "Unnamed: 0                                 \n",
              "0                  K      K  rs2124486056  \n",
              "2                  S      G  rs1223685980  \n",
              "6                  N      S      CD095255  \n",
              "15                 K      R   rs375342012  \n",
              "17                 Y      C   rs104894973  \n",
              "24                 N      D      CM136852  \n",
              "26                 K      K  rs2124486056  \n",
              "27                 K      R      CM920650  \n",
              "28                 K      E  rs2124486060  \n",
              "32                 E      G     CM1210299  \n",
              "33                 S      S  rs2124486161  \n",
              "41                 T      A  rs2124486396  \n",
              "50                 K      R      CM083165  \n",
              "73                 E      G     CM1210299  \n",
              "86                 F      L  rs1459783583  \n",
              "101                I      T   rs104894968  \n",
              "104                I      T   rs104894968  \n",
              "106                L      S      CM117656  \n",
              "107                L      L  rs1603308290  \n",
              "113                F      S   rs104894956  \n",
              "121                Y      Y  rs2124485906  \n",
              "131                N      N      CM144467  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-721ac6ae-ed08-45f5-933b-e4be21a44677\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Hugo_Symbol</th>\n",
              "      <th>gRNA_Target_Sequence</th>\n",
              "      <th>most_severe_consequence</th>\n",
              "      <th>Edited_AA</th>\n",
              "      <th>New_AA</th>\n",
              "      <th>clinical_id</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AAAATGGCCATTCTTCCAGG</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>K</td>\n",
              "      <td>K</td>\n",
              "      <td>rs2124486056</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>S</td>\n",
              "      <td>G</td>\n",
              "      <td>rs1223685980</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AAGAGAATATTCCCGCTCTC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>N</td>\n",
              "      <td>S</td>\n",
              "      <td>CD095255</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>15</th>\n",
              "      <td>SRY</td>\n",
              "      <td>ATTATAAGTATCGACCTCGT</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>K</td>\n",
              "      <td>R</td>\n",
              "      <td>rs375342012</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>17</th>\n",
              "      <td>SRY</td>\n",
              "      <td>ATTATAAGTATCGACCTCGT</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>Y</td>\n",
              "      <td>C</td>\n",
              "      <td>rs104894973</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>24</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CCATGAACGCATTCATCGTG</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>N</td>\n",
              "      <td>D</td>\n",
              "      <td>CM136852</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>26</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CGAAAAATGGCCATTCTTCC</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>K</td>\n",
              "      <td>K</td>\n",
              "      <td>rs2124486056</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>27</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CGAAAAATGGCCATTCTTCC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>K</td>\n",
              "      <td>R</td>\n",
              "      <td>CM920650</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>28</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CGAAAAATGGCCATTCTTCC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>K</td>\n",
              "      <td>E</td>\n",
              "      <td>rs2124486060</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>32</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CTCAGAGATCAGCAAGCAGC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>E</td>\n",
              "      <td>G</td>\n",
              "      <td>CM1210299</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>33</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CTCAGAGATCAGCAAGCAGC</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>S</td>\n",
              "      <td>S</td>\n",
              "      <td>rs2124486161</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>41</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GAAACGGGAGAAAACAGTAA</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "      <td>rs2124486396</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>50</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GCAAGCAGCTGGGATACCAG</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>K</td>\n",
              "      <td>R</td>\n",
              "      <td>CM083165</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>73</th>\n",
              "      <td>SRY</td>\n",
              "      <td>TCAGAGATCAGCAAGCAGCT</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>E</td>\n",
              "      <td>G</td>\n",
              "      <td>CM1210299</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>86</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AAAGGAAGGAAGAGCTTCTC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>F</td>\n",
              "      <td>L</td>\n",
              "      <td>rs1459783583</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>101</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CACACGATGAATGCGTTCAT</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>I</td>\n",
              "      <td>T</td>\n",
              "      <td>rs104894968</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>104</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CCACACGATGAATGCGTTCA</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>I</td>\n",
              "      <td>T</td>\n",
              "      <td>rs104894968</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>106</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CCTGTAATTTCTGTGCCTCC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>L</td>\n",
              "      <td>S</td>\n",
              "      <td>CM117656</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>107</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CCTGTAATTTCTGTGCCTCC</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>L</td>\n",
              "      <td>L</td>\n",
              "      <td>rs1603308290</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>113</th>\n",
              "      <td>SRY</td>\n",
              "      <td>CTGGAAGAATGGCCATTTTT</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>F</td>\n",
              "      <td>S</td>\n",
              "      <td>rs104894956</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>121</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GAGGTCGATACTTATAATTC</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>Y</td>\n",
              "      <td>Y</td>\n",
              "      <td>rs2124485906</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>131</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GGAATATTCTCTTGCACAGC</td>\n",
              "      <td>synonymous_variant</td>\n",
              "      <td>N</td>\n",
              "      <td>N</td>\n",
              "      <td>CM144467</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-721ac6ae-ed08-45f5-933b-e4be21a44677')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-721ac6ae-ed08-45f5-933b-e4be21a44677 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-721ac6ae-ed08-45f5-933b-e4be21a44677');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-2fce45f9-dd07-4f6d-a544-636cb9dbccfd\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-2fce45f9-dd07-4f6d-a544-636cb9dbccfd')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-2fce45f9-dd07-4f6d-a544-636cb9dbccfd button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"    'Hugo_Symbol', 'gRNA_Target_Sequence', 'most_severe_consequence', 'Edited_AA','New_AA', 'clinical_id']]\",\n  \"rows\": 22,\n  \"fields\": [\n    {\n      \"column\": \"Unnamed: 0\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 43,\n        \"min\": 0,\n        \"max\": 131,\n        \"num_unique_values\": 22,\n        \"samples\": [\n          0,\n          73,\n          28\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Hugo_Symbol\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"SRY\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"gRNA_Target_Sequence\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 17,\n        \"samples\": [\n          \"AAAATGGCCATTCTTCCAGG\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"most_severe_consequence\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2,\n        \"samples\": [\n          \"missense_variant\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Edited_AA\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 9,\n        \"samples\": [\n          \"I\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"New_AA\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 12,\n        \"samples\": [\n          \"Y\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"clinical_id\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 19,\n        \"samples\": [\n          \"rs2124486056\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 63
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "protein_df[protein_df.most_severe_consequence == \"missense_variant\"][[\n",
        "    'Hugo_Symbol', 'gRNA_Target_Sequence', 'most_severe_consequence', 'Protein_Position','Protein_Change','curated_Domain']]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 527
        },
        "id": "YTMlWfbEgkwi",
        "outputId": "c2f9864c-9e4c-4677-f692-5b48fd194a9a"
      },
      "id": "YTMlWfbEgkwi",
      "execution_count": 64,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "           Hugo_Symbol  gRNA_Target_Sequence most_severe_consequence  \\\n",
              "Unnamed: 0                                                             \n",
              "1                  SRY  AAACAGTAAAGGCAACGTCC        missense_variant   \n",
              "2                  SRY  AAACAGTAAAGGCAACGTCC        missense_variant   \n",
              "3                  SRY  AAACAGTAAAGGCAACGTCC        missense_variant   \n",
              "4                  SRY  AACGGGACCGCTACAGCCAC        missense_variant   \n",
              "5                  SRY  AACTGGACAACAGGTTGTAC        missense_variant   \n",
              "...                ...                   ...                     ...   \n",
              "130                SRY  GGAATATTCTCTTGCACAGC        missense_variant   \n",
              "132                SRY  GGAATATTCTCTTGCACAGC        missense_variant   \n",
              "136                SRY  GGTGAGCTGGCTGCGTTGAT        missense_variant   \n",
              "142                SRY  GTGCTCCATTCTTGAGTGTG        missense_variant   \n",
              "146                SRY  TCGCTGCAGAGTACCGAAGC        missense_variant   \n",
              "\n",
              "           Protein_Position Protein_Change  \\\n",
              "Unnamed: 0                                   \n",
              "1                        53            K/E   \n",
              "2                        52            S/G   \n",
              "3                     52;53          SK/GE   \n",
              "4                       196            D/G   \n",
              "5                       160            D/G   \n",
              "...                     ...            ...   \n",
              "130                      25            I/T   \n",
              "132                      25          NI/NT   \n",
              "136                     191            S/P   \n",
              "142                     176            M/T   \n",
              "146                     154            C/R   \n",
              "\n",
              "                                               curated_Domain  \n",
              "Unnamed: 0                                                     \n",
              "1                                                         NaN  \n",
              "2                                                         NaN  \n",
              "3                                                         NaN  \n",
              "4                                                  Disordered  \n",
              "5                                                         NaN  \n",
              "...                                                       ...  \n",
              "130                                                       NaN  \n",
              "132                                                       NaN  \n",
              "136                                                Disordered  \n",
              "142                                                Disordered  \n",
              "146         Necessary for interaction with ZNF208 isoform ...  \n",
              "\n",
              "[97 rows x 6 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-1f8646b6-aecf-47fc-bcaa-7ea46798c544\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Hugo_Symbol</th>\n",
              "      <th>gRNA_Target_Sequence</th>\n",
              "      <th>most_severe_consequence</th>\n",
              "      <th>Protein_Position</th>\n",
              "      <th>Protein_Change</th>\n",
              "      <th>curated_Domain</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>53</td>\n",
              "      <td>K/E</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>52</td>\n",
              "      <td>S/G</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AAACAGTAAAGGCAACGTCC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>52;53</td>\n",
              "      <td>SK/GE</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AACGGGACCGCTACAGCCAC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>196</td>\n",
              "      <td>D/G</td>\n",
              "      <td>Disordered</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5</th>\n",
              "      <td>SRY</td>\n",
              "      <td>AACTGGACAACAGGTTGTAC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>160</td>\n",
              "      <td>D/G</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>130</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GGAATATTCTCTTGCACAGC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>25</td>\n",
              "      <td>I/T</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>132</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GGAATATTCTCTTGCACAGC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>25</td>\n",
              "      <td>NI/NT</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>136</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GGTGAGCTGGCTGCGTTGAT</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>191</td>\n",
              "      <td>S/P</td>\n",
              "      <td>Disordered</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>142</th>\n",
              "      <td>SRY</td>\n",
              "      <td>GTGCTCCATTCTTGAGTGTG</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>176</td>\n",
              "      <td>M/T</td>\n",
              "      <td>Disordered</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>146</th>\n",
              "      <td>SRY</td>\n",
              "      <td>TCGCTGCAGAGTACCGAAGC</td>\n",
              "      <td>missense_variant</td>\n",
              "      <td>154</td>\n",
              "      <td>C/R</td>\n",
              "      <td>Necessary for interaction with ZNF208 isoform ...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>97 rows √ó 6 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-1f8646b6-aecf-47fc-bcaa-7ea46798c544')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-1f8646b6-aecf-47fc-bcaa-7ea46798c544 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-1f8646b6-aecf-47fc-bcaa-7ea46798c544');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-0e581230-2675-4ec9-940e-df5af2f9f154\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-0e581230-2675-4ec9-940e-df5af2f9f154')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-0e581230-2675-4ec9-940e-df5af2f9f154 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"    'Hugo_Symbol', 'gRNA_Target_Sequence', 'most_severe_consequence', 'Protein_Position','Protein_Change','curated_Domain']]\",\n  \"rows\": 97,\n  \"fields\": [\n    {\n      \"column\": \"Unnamed: 0\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": 41,\n        \"min\": 1,\n        \"max\": 146,\n        \"num_unique_values\": 97,\n        \"samples\": [\n          87,\n          50,\n          132\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Hugo_Symbol\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"SRY\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"gRNA_Target_Sequence\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 47,\n        \"samples\": [\n          \"AAAGGAAGGAAGAGCTTCTC\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"most_severe_consequence\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"missense_variant\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Protein_Position\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 61,\n        \"samples\": [\n          \"53\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Protein_Change\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 44,\n        \"samples\": [\n          \"IV/TA\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"curated_Domain\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 6,\n        \"samples\": [\n          \"Disordered\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 64
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# When you run BEstimate with off targets\n",
        "grna_df = pandas.read_csv(\"/content/SRY_ABE_NGG_ot_annotated_summary_df.csv\",)\n",
        "grna_df[:5]"
      ],
      "metadata": {
        "id": "_wNKgtQBgk0O"
      },
      "id": "_wNKgtQBgk0O",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Find gRNAs without any off targets\n",
        "grna_df[(grna_df.exact == 1) & (grna_df.mm1 == 0) & (grna_df.mm2 == 0) & (grna_df.mm3 == 0)]"
      ],
      "metadata": {
        "id": "IKo0EdfWofgx"
      },
      "id": "IKo0EdfWofgx",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Sickle cell reversion results"
      ],
      "metadata": {
        "id": "0cb-x4_WMQdB"
      },
      "id": "0cb-x4_WMQdB"
    },
    {
      "cell_type": "code",
      "source": [
        "hbb_mut_df = pandas.read_csv(\"/content/output/HBB_variant_specific_ABE_NGN_edit_df.csv\", index_col=0)\n",
        "hbb_mut_df[:5]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 362
        },
        "id": "w1_BTfivMVIi",
        "outputId": "c382f785-0553-420e-e229-a90a14004a22"
      },
      "id": "w1_BTfivMVIi",
      "execution_count": 16,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                 CRISPR_PAM_Sequence  gRNA_Target_Sequence  \\\n",
              "Hugo_Symbol                                                  \n",
              "HBB          ACCATTGGAAAAGCAACCCCTGC  ACCATTGGAAAAGCAACCCC   \n",
              "HBB          TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
              "HBB          TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
              "HBB          TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
              "HBB          TGGAAAAGCAACCCCTGCCTTGA  TGGAAAAGCAACCCCTGCCT   \n",
              "\n",
              "                       Location  Edit_Location Direction  Strand  \\\n",
              "Hugo_Symbol                                                        \n",
              "HBB          11:5229377-5229399        5229391      left      -1   \n",
              "HBB          11:5229372-5229394        5229391      left      -1   \n",
              "HBB          11:5229372-5229394        5229390      left      -1   \n",
              "HBB          11:5229372-5229394        5229389      left      -1   \n",
              "HBB          11:5229372-5229394        5229388      left      -1   \n",
              "\n",
              "                     Gene_ID Transcript_ID Exon_ID  guide_in_CDS  \\\n",
              "Hugo_Symbol                                                        \n",
              "HBB          ENSG00000244734           NaN     NaN         False   \n",
              "HBB          ENSG00000244734           NaN     NaN         False   \n",
              "HBB          ENSG00000244734           NaN     NaN         False   \n",
              "HBB          ENSG00000244734           NaN     NaN         False   \n",
              "HBB          ENSG00000244734           NaN     NaN         False   \n",
              "\n",
              "             gRNA_flanking_sequences  Edit_in_Exon  Edit_in_CDS   GC%  \\\n",
              "Hugo_Symbol                                                             \n",
              "HBB                              NaN         False        False  50.0   \n",
              "HBB                              NaN         False        False  55.0   \n",
              "HBB                              NaN         False        False  55.0   \n",
              "HBB                              NaN         False        False  55.0   \n",
              "HBB                              NaN         False        False  55.0   \n",
              "\n",
              "             # Edits/guide  Poly_T  mutation_on_guide  guide_change_mutation  \\\n",
              "Hugo_Symbol                                                                    \n",
              "HBB                      1   False              False                  False   \n",
              "HBB                      4   False              False                  False   \n",
              "HBB                      4   False              False                  False   \n",
              "HBB                      4   False              False                  False   \n",
              "HBB                      4   False              False                  False   \n",
              "\n",
              "             mutation_on_window  mutation_on_PAM  \n",
              "Hugo_Symbol                                       \n",
              "HBB                       False            False  \n",
              "HBB                       False            False  \n",
              "HBB                       False            False  \n",
              "HBB                       False            False  \n",
              "HBB                       False            False  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ec4aa5f5-5204-463f-9b20-cfb84c457781\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>CRISPR_PAM_Sequence</th>\n",
              "      <th>gRNA_Target_Sequence</th>\n",
              "      <th>Location</th>\n",
              "      <th>Edit_Location</th>\n",
              "      <th>Direction</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>Transcript_ID</th>\n",
              "      <th>Exon_ID</th>\n",
              "      <th>guide_in_CDS</th>\n",
              "      <th>gRNA_flanking_sequences</th>\n",
              "      <th>Edit_in_Exon</th>\n",
              "      <th>Edit_in_CDS</th>\n",
              "      <th>GC%</th>\n",
              "      <th># Edits/guide</th>\n",
              "      <th>Poly_T</th>\n",
              "      <th>mutation_on_guide</th>\n",
              "      <th>guide_change_mutation</th>\n",
              "      <th>mutation_on_window</th>\n",
              "      <th>mutation_on_PAM</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Hugo_Symbol</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>HBB</th>\n",
              "      <td>ACCATTGGAAAAGCAACCCCTGC</td>\n",
              "      <td>ACCATTGGAAAAGCAACCCC</td>\n",
              "      <td>11:5229377-5229399</td>\n",
              "      <td>5229391</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000244734</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>HBB</th>\n",
              "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
              "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
              "      <td>11:5229372-5229394</td>\n",
              "      <td>5229391</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000244734</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>55.0</td>\n",
              "      <td>4</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>HBB</th>\n",
              "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
              "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
              "      <td>11:5229372-5229394</td>\n",
              "      <td>5229390</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000244734</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>55.0</td>\n",
              "      <td>4</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>HBB</th>\n",
              "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
              "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
              "      <td>11:5229372-5229394</td>\n",
              "      <td>5229389</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000244734</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>55.0</td>\n",
              "      <td>4</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>HBB</th>\n",
              "      <td>TGGAAAAGCAACCCCTGCCTTGA</td>\n",
              "      <td>TGGAAAAGCAACCCCTGCCT</td>\n",
              "      <td>11:5229372-5229394</td>\n",
              "      <td>5229388</td>\n",
              "      <td>left</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000244734</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>55.0</td>\n",
              "      <td>4</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ec4aa5f5-5204-463f-9b20-cfb84c457781')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ec4aa5f5-5204-463f-9b20-cfb84c457781 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ec4aa5f5-5204-463f-9b20-cfb84c457781');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-2fa4f77e-a44d-45f3-8cc9-b145ff7bf6e8\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-2fa4f77e-a44d-45f3-8cc9-b145ff7bf6e8')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-2fa4f77e-a44d-45f3-8cc9-b145ff7bf6e8 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "repr_error": "Out of range float values are not JSON compliant: nan"
            }
          },
          "metadata": {},
          "execution_count": 16
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Find the gRNA changing the variation\n",
        "hbb_mut_df[hbb_mut_df.guide_change_mutation]"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 167
        },
        "id": "z0El5MhHsyre",
        "outputId": "05821ef4-29ad-4f2d-c9a2-1a7183d4c0cf"
      },
      "id": "z0El5MhHsyre",
      "execution_count": 17,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "                 CRISPR_PAM_Sequence  gRNA_Target_Sequence  \\\n",
              "Hugo_Symbol                                                  \n",
              "HBB          ACTTCTCCACAGGAGTCAGATGC  ACTTCTCCACAGGAGTCAGA   \n",
              "\n",
              "                       Location  Edit_Location Direction  Strand  \\\n",
              "Hugo_Symbol                                                        \n",
              "HBB          11:5226994-5227016        5227002     right      -1   \n",
              "\n",
              "                     Gene_ID    Transcript_ID          Exon_ID  guide_in_CDS  \\\n",
              "Hugo_Symbol                                                                    \n",
              "HBB          ENSG00000244734  ENST00000335295  ENSE00001829867          True   \n",
              "\n",
              "             gRNA_flanking_sequences  Edit_in_Exon  Edit_in_CDS   GC%  \\\n",
              "Hugo_Symbol                                                             \n",
              "HBB                              NaN          True         True  50.0   \n",
              "\n",
              "             # Edits/guide  Poly_T  mutation_on_guide  guide_change_mutation  \\\n",
              "Hugo_Symbol                                                                    \n",
              "HBB                      1   False               True                   True   \n",
              "\n",
              "             mutation_on_window  mutation_on_PAM  \n",
              "Hugo_Symbol                                       \n",
              "HBB                        True            False  "
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-096eca9a-a31a-4791-89de-924a622194a8\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>CRISPR_PAM_Sequence</th>\n",
              "      <th>gRNA_Target_Sequence</th>\n",
              "      <th>Location</th>\n",
              "      <th>Edit_Location</th>\n",
              "      <th>Direction</th>\n",
              "      <th>Strand</th>\n",
              "      <th>Gene_ID</th>\n",
              "      <th>Transcript_ID</th>\n",
              "      <th>Exon_ID</th>\n",
              "      <th>guide_in_CDS</th>\n",
              "      <th>gRNA_flanking_sequences</th>\n",
              "      <th>Edit_in_Exon</th>\n",
              "      <th>Edit_in_CDS</th>\n",
              "      <th>GC%</th>\n",
              "      <th># Edits/guide</th>\n",
              "      <th>Poly_T</th>\n",
              "      <th>mutation_on_guide</th>\n",
              "      <th>guide_change_mutation</th>\n",
              "      <th>mutation_on_window</th>\n",
              "      <th>mutation_on_PAM</th>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>Hugo_Symbol</th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "      <th></th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>HBB</th>\n",
              "      <td>ACTTCTCCACAGGAGTCAGATGC</td>\n",
              "      <td>ACTTCTCCACAGGAGTCAGA</td>\n",
              "      <td>11:5226994-5227016</td>\n",
              "      <td>5227002</td>\n",
              "      <td>right</td>\n",
              "      <td>-1</td>\n",
              "      <td>ENSG00000244734</td>\n",
              "      <td>ENST00000335295</td>\n",
              "      <td>ENSE00001829867</td>\n",
              "      <td>True</td>\n",
              "      <td>NaN</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>50.0</td>\n",
              "      <td>1</td>\n",
              "      <td>False</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>True</td>\n",
              "      <td>False</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-096eca9a-a31a-4791-89de-924a622194a8')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-096eca9a-a31a-4791-89de-924a622194a8 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-096eca9a-a31a-4791-89de-924a622194a8');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"hbb_mut_df[hbb_mut_df\",\n  \"rows\": 1,\n  \"fields\": [\n    {\n      \"column\": \"Hugo_Symbol\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"HBB\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"CRISPR_PAM_Sequence\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"ACTTCTCCACAGGAGTCAGATGC\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"gRNA_Target_Sequence\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"ACTTCTCCACAGGAGTCAGA\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Location\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"11:5226994-5227016\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Edit_Location\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 5227002,\n        \"max\": 5227002,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          5227002\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Direction\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"right\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Strand\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": -1,\n        \"max\": -1,\n        \"num_unique_values\": 1,\n        \"samples\": [\n          -1\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Gene_ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"ENSG00000244734\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Transcript_ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"ENST00000335295\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Exon_ID\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          \"ENSE00001829867\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"guide_in_CDS\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [\n          true\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"gRNA_flanking_sequences\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": null,\n        \"max\": null,\n        \"num_unique_values\": 0,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Edit_in_Exon\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Edit_in_CDS\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"GC%\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 50.0,\n        \"max\": 50.0,\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"# Edits/guide\",\n      \"properties\": {\n        \"dtype\": \"number\",\n        \"std\": null,\n        \"min\": 1,\n        \"max\": 1,\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Poly_T\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"mutation_on_guide\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"guide_change_mutation\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"mutation_on_window\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"mutation_on_PAM\",\n      \"properties\": {\n        \"dtype\": \"boolean\",\n        \"num_unique_values\": 1,\n        \"samples\": [],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**Important: WT codon is GAG and the mutant one is GTG\n",
        "Mutation is on 5227002 and mutant sequence between 5227001-5227003 is CAC --> ABE --> CGC (+1 strand)\n",
        "GCG --> Ala --> naturally occurring, non-sickling variant hemoglobin \"Makassar\" (HbG)**"
      ],
      "metadata": {
        "id": "apuL0vGkVjfs"
      },
      "id": "apuL0vGkVjfs"
    },
    {
      "cell_type": "markdown",
      "source": [
        "### Key points to review in your output tables"
      ],
      "metadata": {
        "id": "qzhEu5eMV0QS"
      },
      "id": "qzhEu5eMV0QS"
    },
    {
      "cell_type": "markdown",
      "id": "61c123b1",
      "metadata": {
        "id": "61c123b1"
      },
      "source": [
        "\n",
        "\n",
        "- **Base Change**: Given your experiments of interest, you may highlight gRNA targeting specific domains, post translational modification sites, splice sites or clinically important locations.\n",
        "  - gRNAs that target coding regions resulting in functional consequences like amino acid changes. You can eliminate gRNAs only generating synonymous alterations.\n",
        "  - gRNAs can also edit non-coding regions, you may want to work with a regulatory region, such as promoters, splice sites. (*If unintended, avoid gRNAs that disrupt known splice sites unless this is the intended effect.*)\n",
        "  - gRNAs can replicate or revert known pathogenic SNPs, you may want to investigate disease models or corrections.\n",
        "  - gRNAs targeting highly conserved sequences tend to have more severe functional consequences. You can check the functional consequences and select gRNAs of your interest.\n",
        "\n",
        "- **Off-targets**: It is a good practice to choose gRNAs with minimal off-target effects.\n",
        "\n",
        "Note: On-Target Efficiency: You may want to select gRNAs with a high on-target efficiency which can you obtain through BE-Hive. (*It is not provided by BEstimate*)\n"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## 5Ô∏è‚É£ Controls in library design\n"
      ],
      "metadata": {
        "id": "ggOvVhDhnUja"
      },
      "id": "ggOvVhDhnUja"
    },
    {
      "cell_type": "markdown",
      "source": [
        "When generating a gRNA library for base editing, incorporating proper controls is essential for ensuring the reliability and interpretability of your experimental results. The controls help validate the functional outcomes of your gRNAs.\n",
        "\n",
        "1. Positive Controls which help confirm that your base editing system is working efficiently and that the experimental conditions are optimal.\n",
        "\n",
        "  - gRNAs targeting essential genes that are essential for cell viability (such as housekeeping genes) where editing should have measurable phenotypic effects like cell death or reduced growth.\n",
        "\n",
        "2. Negative Controls which are critical to assess the background levels of editing and off-target effects. They ensure that observed changes are due to base editing rather than random or non-specific effects.\n",
        "\n",
        "  - Non-targeting gRNAs help establish the baseline for off-target activity and general effects of transfection or editing. These controls are typically random sequences with no homology to the genome but are designed to resemble real gRNAs in structure.\n",
        "\n",
        "  - gRNAs targeting non-essential genes that are expected to result in no significant phenotypic effect from the base editing.\n",
        "\n"
      ],
      "metadata": {
        "id": "ZVtK60bznWnA"
      },
      "id": "ZVtK60bznWnA"
    },
    {
      "cell_type": "markdown",
      "id": "3c50a40e",
      "metadata": {
        "id": "3c50a40e"
      },
      "source": [
        "\n",
        "## üõ†Ô∏è Troubleshooting Tips\n",
        "\n",
        "‚ùó **No module named BEstimate** ‚Üí Rerun the installation cell at the top and do not forget to restart the session! Then you should not run it again.\n",
        "\n",
        "\n",
        "‚ùó **Permission errors** ‚Üí Make sure you‚Äôre running in a writable Colab notebook.  "
      ]
    },
    {
      "cell_type": "markdown",
      "id": "cff10b7f",
      "metadata": {
        "id": "cff10b7f"
      },
      "source": [
        "\n",
        "## üéâ Wrap-up\n",
        "\n",
        "With this practical course, you have now:\n",
        "\n",
        "‚úÖ Set up your environment\n",
        "\n",
        "‚úÖ Designed base editor gRNAs with BEstimate\n",
        "\n",
        "‚úÖ Learned how to interpret your results\n",
        "\n",
        "‚úÖ Learned things to consider while selecting your gRNAs and designing your library\n",
        "\n",
        "\n",
        "\n",
        "**Next steps:** You can try using your own genes or variants as input!\n",
        "\n",
        "\n",
        "\n",
        "Questions? Ask during the live session or contact me at cd7@sanger.ac.uk\n"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "kotm4Xftsl_w",
        "PpI6itIMNF10"
      ]
    },
    "language_info": {
      "name": "python"
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}